pkg/lib/oauth/handler/handler_token.go

```diff
@@ -559,20 +559,25 @@ func (h *TokenHandler) resolveIDTokenSession(idToken jwt.Token) (sidSession sess
 	return sidSession, true, nil
 }
 
-func (h *TokenHandler) verifyIDTokenDeviceSecretHash(idToken jwt.Token, deviceSecret string) error {
+func (h *TokenHandler) verifyIDTokenDeviceSecretHash(offlineGrant *oauth.OfflineGrant, idToken jwt.Token, deviceSecret string) error {
+	// Always do all checks to ensure this method consumes constant time
+	var err error = nil
 	deviceSecretHash := oauth.HashToken(deviceSecret)
 	dsHashInterface, ok := idToken.Get(string(model.ClaimDeviceSecretHash))
 	if !ok {
-		return fmt.Errorf("ds_hash does not exist")
+		err = fmt.Errorf("ds_hash does not exist")
```

Comment:
`err = protocol.NewError("invalid_grant", "expected ds_hash to be present in id token (subject_token)"`

Reply:
These rules are documented in the spec, which is a public document. It is OK to return more detailed error messages.